fix: services/httpd: parse correctly Accept header with extra test cases

jsign opened the pull request:
Closes #14296

Until the standard lib decides to include a standard function to correctly parse Accept headers, we can rely on [goautoneg](https://github.com/golang/go/issues/19307#issuecomment-511133691) library which is a Github port used by Prometheus.

I also added sub-tests as Table Tests for multiples scenarios around this problem.

Adding a new dependency for this amount of code may hurt a bit, so maybe we could add the code directly too.

- [ ] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] Rebased/mergeable
- [x] Tests pass
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

jsign commented:
The failed test case seems to be a [reported flaky test](https://github.com/influxdata/influxdb/issues/11531)

jsign commented:
@russorat , I rebased it to `1.8` to keep it clean. 
Is `1.8` OK for merging or do you prefer on `master`?

jsign commented:
@jsternberg, pushed changes in the direction you mentioned.

jsternberg commented on services/httpd/response_writer.go:
```diff
@@ -25,19 +25,27 @@ type ResponseWriter interface {
-		w.Header().Add("Content-Type", "application/json")
-		rw.formatter = &jsonFormatter{Pretty: pretty}
+
+	acceptHeaders := parseAccept(r.Header.Get("Accept"))
```
Can you change this to `r.Header` as the argument and change `parseAccept` to take a `[]string` value? I think it's possible for someone to do this:

```
Accept: application/csv
Accept: application/x-msgpack
```

jsternberg commented on services/httpd/accept.go:
```diff
@@ -0,0 +1,68 @@
+/* This file is a copy of github.com/markusthoemmes/goautoneg
```
Can you copy the full license into this file? Something like:

```
// This file is an adaptation of https://github.com/markusthoemmes/goautoneg.
// The copyright and license header are reproduced below.
//
// (copy of the license)
//
// (link to the license)
```

jsternberg commented on services/httpd/response_writer.go:
```diff
@@ -25,19 +25,27 @@ type ResponseWriter interface {
+
+	acceptHeaders := parseAccept(r.Header.Get("Accept"))
+	for _, accept := range acceptHeaders {
+		switch accept.Type + "/" + accept.SubType {
```
It seems to me this code handles wildcards incorrectly. If I am understanding the spec correctly, a `*` would match any one of these. Is this correct? If it is, I have some recommendations for how we can change the code. If it isn't, then we can skip past that.

jsign replied:
Uhm, I agree. Makes sense.

jsternberg commented on services/httpd/response_writer_test.go:
```diff
@@ -19,63 +19,77 @@ import (
-	r := &http.Request{
-		Header: header,
-		URL:    &url.URL{},
+	tableTest := []struct {
```
I'd like some wildcard tests here.

jsign commented:
@jsternberg, I applied your suggestions, and also switched to performance mode since the code had many allocations that could be avoided. And considering this is a hot-path, deserved some extra attention.

Some points about the changes (apart from your comments):
- make `csvFormatter` stateless
- all known content types and formatters are defined as vars, so to a ton of allocations. This is the reason why I made `csvFormatter` stateless (since it wasn't).
- avoided string concats to save extra allocations (that's why `full` exist apart from splited parts)
- added `*` test case
- in case of `*/*` the switch isn't biased to a definite option. This is strictly correct (since any option complies to request), but maybe another default is wanted.
- as can be seen, adding future content types is much less verbose

jsternberg commented:
Sorry I haven't had time to review this again. I will review by end of day tomorrow.

dgnorton commented:
@jsternberg can you review this for the 1.8 release coming up?

jsternberg commented:
I reverted the changes to the formatters and modified the creation of those formatters to a factory method. I think that the state is actually used by the code so a global instance doesn't work, but I also don't remember well enough to know for certain so I don't want those to change.

I am going to wait for the tests and run some myself and then merge this.
